// @aws-amplify/data-schema/dist/cjs/CustomOperation.js
'use strict';

Object.defineProperty(exports, "__esModule", { value: true });
exports.subscription = exports.mutation = exports.query = exports.CustomOperationNames = void 0;
const util_1 = require("./util");
const Authorization_1 = require("./Authorization");
const queryBrand = 'queryCustomOperation';
const mutationBrand = 'mutationCustomOperation';
const subscriptionBrand = 'subscriptionCustomOperation';
exports.CustomOperationNames = [
    'Query',
    'Mutation',
    'Subscription',
];
function brandedBuilder(builder, brandValue) {
    return { ...builder, ...(0, util_1.brand)(brandValue) };
}
function _custom(typeName, brand) {
    const data = {
        arguments: {},
        returnType: null,
        functionRef: null,
        authorization: [],
        typeName: typeName,
        handlers: null,
        subscriptionSource: [],
    };
    const builder = brandedBuilder({
        arguments(args) {
            data.arguments = args;
            return this;
        },
        returns(returnType) {
            data.returnType = returnType;
            return this;
        },
        function(functionRefOrName) {
            data.functionRef = functionRefOrName;
            return this;
        },
        authorization(callback) {
            const rules = callback(Authorization_1.allowForCustomOperations);
            data.authorization = Array.isArray(rules) ? rules : [rules];
            return this;
        },
        handler(handlers) {
            data.handlers = Array.isArray(handlers)
                ? handlers
                : [handlers];
            return this;
        },
        for(source) {
            data.subscriptionSource = Array.isArray(source) ? source : [source];
            return this;
        },
    }, brand);
    return { ...builder, data };
}
function query() {
    return _custom('Query', queryBrand);
}
exports.query = query;
function mutation() {
    return _custom('Mutation', mutationBrand);
}
exports.mutation = mutation;
function subscription() {
    return _custom('Subscription', subscriptionBrand);
}
exports.subscription = subscription;
