// @aws-amplify/data-schema/dist/cjs/Handler.js
'use strict';

Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.getHandlerData = void 0;
const util_1 = require("./util");
const dataSymbol = Symbol('Data');
function buildHandler(brandName) {
    return (0, util_1.brand)(brandName);
}
function getHandlerData(handler) {
    return handler[dataSymbol];
}
exports.getHandlerData = getHandlerData;
//#region handler.inlineSql
const inlineSqlBrand = 'inlineSql';
function inlineSql(sql) {
    return { [dataSymbol]: sql, ...buildHandler(inlineSqlBrand) };
}
//#endregion
//#region handler.sqlReference
const sqlReferenceBrand = 'sqlReference';
function sqlReference(sqlFilePath) {
    // used to determine caller directory in order to resolve relative path downstream
    const stack = new Error().stack;
    return {
        [dataSymbol]: { stack, entry: sqlFilePath },
        ...buildHandler(sqlReferenceBrand),
    };
}
const customHandlerBrand = 'customHandler';
function custom(customHandler) {
    // used to determine caller directory in order to resolve relative path downstream
    const stack = new Error().stack;
    return {
        [dataSymbol]: { ...customHandler, stack },
        ...buildHandler(customHandlerBrand),
    };
}
const functionHandlerBrand = 'functionHandler';
function fcn(fn) {
    return { [dataSymbol]: fn, ...buildHandler(functionHandlerBrand) };
}
//#endregion
exports.handler = {
    inlineSql,
    sqlReference,
    custom,
    function: fcn,
};
